engine/src/io/data_parser/sql/PostgreSQLParser.cpp

```diff
@@ -0,0 +1,549 @@
+/*
+ * Copyright 2021 BlazingDB, Inc.
+ *     Copyright 2021 Cristhian Alberto Gonzales Castillo
+ * <cristhian@blazingdb.com>
+ */
+
+#include <array>
+
+#include <cudf/io/types.hpp>
+#include <cudf/utilities/bit.hpp>
+#include <libpq-fe.h>
+#include <netinet/in.h>
+
+#include "PostgreSQLParser.h"
+#include "sqlcommon.h"
+
+namespace ral {
+namespace io {
+
+static const std::array<const char *, 6> postgresql_string_type_hints = {
+    "character", "character varying", "bytea", "text", "anyarray", "name"};
+
+static inline bool postgresql_is_cudf_string(const std::string &hint) {
+  const auto *result =
+      std::find_if(std::cbegin(postgresql_string_type_hints),
+                   std::cend(postgresql_string_type_hints),
+                   [&hint](const char *c_hint) {
+                     return std::strcmp(c_hint, hint.c_str()) == 0;
+                   });
+  return result != std::cend(postgresql_string_type_hints);
+}
+
+static inline cudf::io::table_with_metadata
+read_postgresql(const std::shared_ptr<PGresult> &pgResult,
+                const std::vector<int> &column_indices,
+                const std::vector<cudf::type_id> &cudf_types,
+                const std::vector<std::size_t> &column_bytes,
+                std::size_t total_rows) {
+  const std::size_t resultNfields = PQnfields(pgResult.get());
+  if (resultNfields != column_indices.size() ||
+      resultNfields != column_bytes.size() ||
+      resultNfields != cudf_types.size()) {
+    throw std::runtime_error(
+        "Not equal columns for indices and bytes in PostgreSQL filter");
+  }
+
+  std::vector<void *> host_cols;
+  host_cols.reserve(resultNfields);
+  const int resultNtuples = PQntuples(pgResult.get());
+  const std::size_t bitmask_allocation =
+      cudf::bitmask_allocation_size_bytes(resultNtuples);
+  const std::size_t num_words = bitmask_allocation / sizeof(cudf::bitmask_type);
+  std::vector<std::vector<cudf::bitmask_type>> null_masks(resultNfields);
+  std::transform(
+      column_indices.cbegin(),
+      column_indices.cend(),
+      std::back_inserter(host_cols),
+      [&pgResult, &cudf_types, &null_masks, num_words, resultNtuples](
+          const int projection_index) {
+        null_masks[projection_index].resize(num_words, 0);
+        const int fsize = PQfsize(pgResult.get(), projection_index);
+        if (fsize < 0) {  // STRING, STRUCT, LIST, and similar cases
+          auto *string_col = new cudf_string_col();
+          string_col->offsets.reserve(resultNtuples + 1);
+          string_col->offsets.push_back(0);
+          return static_cast<void *>(string_col);
+        }
+        // primitives cases
+        const cudf::type_id cudf_type_id = cudf_types[projection_index];
+        switch (cudf_type_id) {
+        case cudf::type_id::INT8: {
```

Comment:
alot of this code seems to be shared with the mysqlparser perhaps the abstract_sql_parser should implement some methods :)

Reply:
i agree

Reply:
yep that was the reason of those abstract classes (common spaces), but I din't have enough time to move the code :p, I'll be move it there cc @gcca

Reply:
I think this refactoring can be done when the postgresql or sqlite implementations are done

Reply:
Great idea!